chain/chain/src/garbage_collection.rs

```diff
@@ -462,6 +478,59 @@ impl<'a> ChainStoreUpdate<'a> {
         shard_uids_to_gc
     }
 
+    /// GC in-memory tries and state after catchup.
+    fn clear_obsolete_shard_data(
+        &mut self,
+        runtime: &dyn RuntimeAdapter,
+        epoch_manager: &dyn EpochManagerAdapter,
+        block_hash: CryptoHash,
+        account_id: &AccountId,
+    ) -> Result<(), Error> {
+        // Check if the last block of the current epoch is completed.
+        if !epoch_manager.is_last_block_in_finished_epoch(&block_hash)? {
+            return Ok(());
+        }
+        let block_info = epoch_manager.get_block_info(&block_hash)?;
+        let first_block_epoch_hash = block_info.epoch_first_block();
+        if first_block_epoch_hash == &CryptoHash::default() {
+            // TODO(#10844) Probably we should GC here as well.
+            return Ok(());
+        }
+        let block_header = self.get_block_header(&block_hash).expect("block header must exist");
+        let current_epoch_shard_layout =
+            epoch_manager.get_shard_layout(block_header.epoch_id()).expect("epoch info must exist");
+        let shard_uids = current_epoch_shard_layout.shard_uids().collect_vec();
+        let mut shards_still_tracked_in_next_epoch = vec![];
+
+        for shard_uid in &shard_uids {
+            if epoch_manager.cares_about_shard_in_epoch(
```

Comment:
I think we may want to use ShardTracker here, which also takes into account voluntary additional shard tracking specified in the config.